Ensure `preprocessTree` / `postprocessTree` are invoked properly for lazy engine styles.

rwjblue opened the pull request:
New / tweaked styles pipeline is:

1. Invoke `preprocessTree` on all of the engines dependencies, with the raw `addon/styles` tree.
2. Process the raw tree through all the preprocessors setup in the registry.
3. Concat `**/*.css` from the result of 2 into `engine.css`
4. Concat vendor styles (dependent addons' `addon-styles` trees) into `engine-vendor.css`
5. Process `app.import`'s related to styles.
6. Invoke `postprocessTree` on all of the engines dependencies, with the final combined output of 3, 4, and 5.

Fixes #322

rwjblue commented:
The "diff tests" are failing due to a mismatched `ember-cli` version when ember-engines is linked within engine-testing.  There is actually no change in output before/after this PR.  The output is exactly the same as [this commit in engine-testing-dist-output](https://github.com/rwjblue/engine-testing-dist-output/tree/bdf714583010dee93e7c00248cee02cf92afc4a6).

stefanpenner commented on lib/engine-addon.js:
```diff
@@ -387,10 +387,76 @@ module.exports = {
+        // This gives the chance for somebody to prepend onto the vendor files.
+        var vendorCSSImportTree = buildVendorCSSWithImports.call(this, concatMergedVendorCSSTree);
+
+        var mergedVendorCSSWithImportAndEngineStylesTree = mergeTrees(
```
do we need https://github.com/ember-cli/ember-cli/blob/master/lib/broccoli/merge-trees.js ?

rwjblue replied:
Yes, I think we do (to help eliminate wasted merges and whatnot).  I'll update.

rwjblue commented on lib/engine-addon.js:
```diff
@@ -387,10 +387,76 @@ module.exports = {
       };
 
+      this.compileLazyEngineStyles = function compileLazyEngineStyles(vendorTree, externalTree) {
+        var vendorCSSTree = buildVendorCSSTree.call(this, vendorTree);
```
@stefanpenner - Yes, they are `memoize`ed with a custom `memoize` util function already.  We can clean it up a bit and make sure they are cached in the same way in future refactors.

rwjblue commented on lib/engine-addon.js:
```diff
@@ -387,10 +387,76 @@ module.exports = {
+        var engineStylesTree = this._treeFor('addon-styles');
+
+        var primaryStyleTree = null;
+        if (engineStylesTree) {
```
> I suspect it may have to do with actually needing to keep vendor + add-on style separate?

Confirm, that is exactly right. If we merge them, then do the `concat` for `**/*.css` we loose the ability to actually have separate `engine.css` and `engine-vendor.css`.

rwjblue commented:
Latest commit bumps `ember-cli` version to match what engine-testing uses, which fixes the node:ci tests (and confirms the built asset output is identical).

trentmwillis commented on lib/engine-addon.js:
```diff
@@ -387,10 +387,76 @@ module.exports = {
+        var engineStylesOutputDir = 'engines-dist/' + this.name + '/assets/';
+
+        // Get base styles tree.
+        var engineStylesTree = this._treeFor('addon-styles');
```
Should we memoize/cache the result of `_treeFor('addon-styles')`? Since it is called in more than one location.

rwjblue replied:
Its only called once per engine instance, the first invocation is for non-lazy engines and the second is for lazy engines.  I think we can definitely swap to memoize though, will update.

rwjblue commented:
@trentmwillis - Should be ready for another look...

alexander-alvarez commented on lib/engine-addon.js:
```diff
@@ -358,14 +361,15 @@ module.exports = {
+        var compiledEngineCSSTree = this.compileStyles(engineCSSTree);
 
-        // If this engine is lazy or any of its parents are lazy we need to
+        // If any of this engines ancestors are lazy we need to
```
`engine's`

alexander-alvarez commented on lib/engine-addon.js:
```diff
@@ -358,14 +361,15 @@ module.exports = {
 
-        // If this engine is lazy or any of its parents are lazy we need to
+        // If any of this engines ancestors are lazy we need to
         // remove its routes, because we've already accounted for our routes
```
`its` being the engine or the parent's ?

nathanhammond commented on lib/engine-addon.js:
```diff
@@ -387,10 +391,76 @@ module.exports = {
+        var primaryStyleTree = null;
+        if (engineStylesTree) {
+          var preprocessedEngineStylesTree;
+          // `_addonPreprocessTree` was added in Ember 2.12 to ensure
```
Ember CLI 2.12.

rwjblue replied:
Seems fine to me, I'll add an `ember-cli-version-checker` `assertAbove` for it.

nathanhammond commented on lib/engine-addon.js:
```diff
@@ -358,14 +361,15 @@ module.exports = {
         // NOT LAZY LOADING!
         // This is the scenario where we want to act like an addon.
-        var engineCSSTree = buildEngineCSSTree.call(this);
+        var engineCSSTree = buildEngineStyleTree.call(this);
```
This is now returning the nested already-preprocessed child CSS trees via https://github.com/ember-cli/ember-cli/blob/16e4492c9ebf3348eb0f31df17215810674dbdf6/lib/models/addon.js#L781-L794

You split it to enable memoization. Can some of this end up documented in comments?

rwjblue replied:
Yep, good point.  Updating with comments...

nathanhammond commented on lib/engine-addon.js:
```diff
@@ -387,10 +391,76 @@ module.exports = {
 
+      this.compileLazyEngineStyles = function compileLazyEngineStyles(vendorTree, externalTree) {
+        var vendorCSSTree = buildVendorCSSTree.call(this, vendorTree);
+        var engineStylesOutputDir = 'engines-dist/' + this.name + '/assets/';
```
I'd prefer `path.join` unless `preprocessCss` expects POSIX in which case `ensure-posix-path`. We currently have to know whether we've included leading/trailing slashes anywhere we modify it such as https://github.com/ember-engines/ember-engines/pull/324/files#diff-1c2c32318ef6ba8d9be4e7b085f8ecb5R419

rwjblue replied:
I don't really think this changes much, I'd prefer to duplicate than pull in a dep and join.

nathanhammond commented on lib/engine-addon.js:
```diff
@@ -406,21 +476,15 @@ module.exports = {
-            outputFile: 'engines-dist/' + this.name + '/assets/engine.css'
-          });
-        }
+        var finalStylesTree = this.compileLazyEngineStyles(vendorTree, externalTree);
```
I'm concerned that this doesn't prevent hoisting of nested lazy engine's styles but haven't traced through it. I believe that this case isn't covered in our fixtures.

Given: `root > lazy1 > lazy2 > eager`

We would expect:
- `root.css` to have only its own assets.
- `lazy1/engine-vendor.css` to be empty.
- `lazy2/engine-vendor.css` to contain `eager`'s styles.

rwjblue replied:
Added that fixture and rebased, https://travis-ci.org/ember-engines/ember-engines/jobs/192898083 shows no diff before and after this change.  

Also, if you are curious, the assets do seem to be appended to the correct place `dist/lazy-in-lazy/assets/engine-vendor.css`.

nathanhammond replied:
Thank you for the new test case; this was my main concern I thought of with this change (though not necessarily stemming from this change).

nathanhammond commented on lib/engine-addon.js:
```diff
@@ -387,10 +391,76 @@ module.exports = {
+          var preprocessedEngineStylesTree;
+          // `_addonPreprocessTree` was added in Ember 2.12 to ensure
+          // addons had preprocessTree / postprocessTree invoked properly
+          // this emulates the normal behavior before and after that version
```
I'd rather we not support both versions. I don't want to accumulate version-specific cruft in a 0.X addon.

nathanhammond replied:
Note this should probably mean that 0.5.0 is blocked on at least 2.12.0-beta.1.

rwjblue replied:
Yep, agreed.  For now, I added a simple feature detect that throws an error if what we need is missing.  I totally agree that we shouldn't release 0.5.0 final until after there is a tagged/released build of ember-cli@2.12 (possibly even until 2.12.0 final is shipped).

nathanhammond commented on lib/engine-addon.js:
```diff
@@ -387,10 +391,76 @@ module.exports = {
+
+        var concatVendorCSSTree = concat(vendorCSSTree, {
+          allowNone: true,
+          inputFiles: ['**/*.css'],
```
😢

rwjblue replied:
😿

rwjblue commented:
Thank you @trentmwillis, @nathanhammond, and @stefanpenner for reviewing and providing feedback.  I _think_ I have addressed all of the concerns at this point, so I am going to go ahead and land this.  If I've missed something please open an issue and I'll jump on it...
